File: yt/yt/flow/examples/python/shuffle/event_mapper.py
"""EventMapper: RowFunction that parses JSON data and emits event messages."""

import json
import logging

from yt.yt.flow.library.python.companion.computation import RowFunction

log = logging.getLogger(__name__)


# [BEGIN event_mapper]
class EventMapper(RowFunction):
    """Parses JSON from data field and emits event messages with key fields."""

    def on_message(self, message, output, ctx):
        data = message.payload["data"]
        try:
            parsed = json.loads(data)
        except (json.JSONDecodeError, TypeError) as e:
            log.error("Error parsing json from data field: %s", e)
            raise

        builder = ctx.message_builder("event")
        builder.set("value", parsed.get("value"))
        builder.set("key_a", parsed.get("key_a"))
        builder.set("key_b", parsed.get("key_b"))
        builder.set("key_c", parsed.get("key_c"))
        builder.set("key_d", parsed.get("key_d"))
        output.add_message(builder.finish())
# [END event_mapper]

File: yt/yt/flow/examples/go/shuffle/event_mapper.go
package main

import (
	"context"
	"encoding/json"
	"fmt"

	"go.ytsaurus.tech/yt/go/flow"
)

const eventStreamID = "event"

// [BEGIN event]

type sourceMessage struct {
	flow.YSONMessage
	Data []byte `yson:"data"`
}

type event struct {
	flow.YSONMessage
	KeyA  uint64 `json:"key_a" yson:"key_a"`
	KeyB  uint64 `json:"key_b" yson:"key_b"`
	KeyC  uint64 `json:"key_c" yson:"key_c"`
	KeyD  uint64 `json:"key_d" yson:"key_d"`
	Value string `json:"value" yson:"value"`
}

// [END event]

// [BEGIN event_mapper]

type eventMapper struct{}

var _ flow.RowFunction = (*eventMapper)(nil)

func (*eventMapper) OnMessage(
	ctx context.Context,
	rt flow.Runtime,
	msg flow.ExtendedMessage,
	out flow.OutputCollector,
) error {
	var input sourceMessage
	if err := msg.ConvertTo(&input); err != nil {
		return err
	}

	parsed := flow.NewYSONMessage[event](eventStreamID)
	if err := json.Unmarshal(input.Data, parsed); err != nil {
		return fmt.Errorf("parsing the data column: %w", err)
	}
	parsed.Meta.StreamID = eventStreamID
	message, err := flow.ConvertFrom(rt, parsed)
	if err != nil {
		return err
	}

	out.AddMessage(message)
	return nil
}

// [END event_mapper]
